api/v1alpha/src/lib/metaManager.js

```diff
@@ -46,25 +48,26 @@ async function performDatasetMetadataUpdate(projectId, datasetId, accounts) {
         let a = metadata.access[i];
         if (a.view && a.view.projectId && a.view.datasetId && a.view.tableId) {
             const _viewExists = await bigqueryUtil.viewExists(a.view.datasetId, a.view.tableId);
-            // If view no longer exists, remove it.
+            // If view no longer exists, remove it
             if (_viewExists === false) {
                 console.log(`Removing authorized view as it does not exist: ${a.view.datasetId}.${a.view.tableId}`);
                 metadata.access.splice(i, 1);
                 isDirty = true;
             }
         }
         else if (a.role === 'READER') {
-            // 3. Remove all 'READER' userByEmail and groupByEmail accounts
-            // Get keys for the current access object.
+            // 3. Remove all 'READER' userByEmail and groupByEmail accounts that should not have access
             const aKeys = Object.keys(a);
             if (aKeys.length === 2) {
                 const accessType = aKeys[1];
                 const accessId = a[accessType];
                 if (accessTypes.includes(accessType)) {
-                    // If we find the record, remove it
-                    console.log(`Deleting user: ${accessType}:${accessId} from datasetId: ${datasetId}`);
-                    metadata.access.splice(i, 1);
-                    isDirty = true;
+                    const shouldHaveAccess = underscore.findWhere(accounts, { email: accessId, emailType: accessType });
```

Comment:
Improvement so that we only remove users that no longer have access to the policy here, rather than always refreshing the whole access metadata.